Source/Renderer/ShaderBuilder.js

```diff
@@ -0,0 +1,280 @@
+import Check from "../Core/Check.js";
+import defined from "../Core/defined.js";
+import defaultValue from "../Core/defaultValue.js";
+import DeveloperError from "../Core/DeveloperError.js";
+import ShaderDestination from "./ShaderDestination.js";
+import ShaderProgram from "./ShaderProgram.js";
+import ShaderSource from "./ShaderSource.js";
+
+/**
+ * An object that makes it easier to build the text of a {@link ShaderProgram}. This tracks GLSL code for both the vertex shader and the fragment shader.
+ * <p>
+ * For vertex shaders, the shader builder tracks a list of <code>#defines</code>,
+ * a list of attributes, a list of uniforms, and a list of shader lines. It also
+ * tracks the location of each attribute so the caller can easily build the {@link VertexArray}
```

Comment:
Hm, the attribute locations are only needed to build the ShaderProgram, but yeah, it doesn't hurt to make the dictionary available to the caller. Plus it might help with unit tests, I'll see.